packages/xstate-vue/src/index.ts

```diff
@@ -1,2 +1,106 @@
-export { useMachine } from './useMachine';
-export { useService } from './useService';
+import { ref, watch, isRef, onBeforeUnmount, Ref } from '@vue/composition-api';
+import {
+  interpret,
+  EventObject,
+  StateMachine,
+  State,
+  Interpreter,
+  InterpreterOptions,
+  MachineOptions,
+  StateConfig
+} from 'xstate';
+
+interface UseMachineOptions<TContext, TEvent extends EventObject> {
+  /**
+   * If provided, will be merged with machine's `context`.
+   */
+  context?: Partial<TContext>;
+  /**
+   * The state to rehydrate the machine to. The machine will
+   * start at this state instead of its `initialState`.
+   */
+  state?: StateConfig<TContext, TEvent>;
+}
+
+export function useMachine<TContext, TEvent extends EventObject>(
+  machine: StateMachine<TContext, any, TEvent>,
+  options: Partial<InterpreterOptions> &
+    Partial<UseMachineOptions<TContext, TEvent>> &
+    Partial<MachineOptions<TContext, TEvent>>
+): {
+  state: Ref<State<TContext, TEvent>>;
+  send: Interpreter<TContext, any, TEvent>['send'];
+  service: Interpreter<TContext, any, TEvent>;
+} {
+  const {
+    context,
+    guards,
+    actions,
+    activities,
+    services,
+    delays,
+    state: rehydratedState,
+    ...interpreterOptions
+  } = options;
+
+  const machineConfig = {
+    context,
+    guards,
+    actions,
+    activities,
+    services,
+    delays
+  };
+
+  const createdMachine = machine.withConfig(machineConfig, {
+    ...machine.context,
+    ...context
+  } as TContext);
+
+  const service = interpret(createdMachine, interpreterOptions).start(
+    rehydratedState ? State.create(rehydratedState) : undefined
+  );
+
+  const state = ref<State<TContext, TEvent>>(service.state);
+
+  service.onTransition(currentState => {
```

Comment:
I still have no idea if this is OK in Vue. Can we set up a subscription in this setup function? Or should this be deferred until some lifecycle hook, like `onBeforeMount`? Do we know some Vue expert that could answer this with confidence?